cedar-policy-core/src/ast/expr.rs

```diff
@@ -1169,23 +1166,30 @@ pub mod expression_construction_errors {
     use smol_str::SmolStr;
     use thiserror::Error;
 
-    /// The same key occurred two or more times in a single record literal
+    /// The same key occurred two or more times
     //
     // CAUTION: this type is publicly exported in `cedar-policy`.
     // Don't make fields `pub`, don't make breaking changes, and use caution
     // when adding public methods.
     #[derive(Debug, PartialEq, Eq, Clone, Diagnostic, Error)]
-    #[error("duplicate key `{key}` in record literal")]
-    pub struct DuplicateKeyInRecordLiteralError {
-        /// The key which occurred two or more times in the record literal
+    #[error("duplicate key `{key}` {context}")]
+    pub struct DuplicateKeyError {
+        /// The key which occurred two or more times
         pub(crate) key: SmolStr,
+        /// Information about where the duplicate key occurred (e.g., "in record literal")
+        pub(crate) context: &'static str,
     }
 
-    impl DuplicateKeyInRecordLiteralError {
-        /// Get the key which occurred two or more times in the record literal
+    impl DuplicateKeyError {
+        /// Get the key which occurred two or more times
         pub fn key(&self) -> &str {
             &self.key
         }
+
+        /// Make a new error with an updated `context` field
+        pub fn set_context(self, context: &'static str) -> Self {
```

Comment:
I'm not sure if this is general idiomaticity guideline or not, but I would expect `set_` to be a method that takes `&mut Self`, and `with_` to be a method that takes `self` and returns `Self`, like `Value::with_maybe_source_loc()` in Core.  Take or leave this advice